lib/elixir/src/elixir_clauses.erl

```diff
@@ -40,8 +40,28 @@ guard({'when', Meta, [Left, Right]}, E) ->
   {ELeft, EL}  = guard(Left, E),
   {ERight, ER} = guard(Right, EL),
   {{'when', Meta, [ELeft, ERight]}, ER};
-guard(Other, E) ->
-  elixir_expand:expand(Other, E).
+guard(Guard, E) ->
+  {EGuard, EG} = elixir_expand:expand(Guard, E),
+  ok = warn_zero_length_guard(EGuard, EG),
+  {EGuard, EG}.
+
+warn_zero_length_guard(
+  {
+    {'.', _, [erlang, '==']},
+    Meta,
+    [{{'.', _, [erlang, length]}, _, [Arg]}, 0]
+  },
+  E
+) ->
+  ArgS = 'Elixir.Macro':to_string(Arg),
+  MessageFmt = "\"length(~ts) == 0\" is discouraged since it has to traverse the whole list to check if it is empty or not. Prefer \"~ts == []\" instead",
+  elixir_errors:warn(?line(Meta), ?key(E, file), io_lib:format(MessageFmt, [ArgS, ArgS])),
```

Comment:
Personally I would restructure this code to make it a bit more readable

```erlang
ArgString = 'Elixir.Macro':to_string(Arg),
Message =
  "\"length(~ts) == 0\" is discouraged since it has to traverse "
  "the whole list to check if it is empty or not. Prefer \"~ts == []\" instead",
elixir_errors:warn(?line(Meta), ?key(E, file), io_lib:format(Message, [ArgS, ArgS]));
```

Wdyt?

Reply:
What about:

```erlang
  ArgString = 'Elixir.Macro':to_string(Arg),
  Message = io_lib:format("\"length(~ts) == 0\" is discouraged since it has to traverse "
                          "the whole list to check if it is empty or not. Prefer "
                          \"~ts == []\" instead", [ArgString, ArgString])
  elixir_errors:warn(?line(Meta), ?key(E, file), Message);
```